common/src/main/java/org/conscrypt/DuckTypedHpkeSpi.java

```diff
@@ -0,0 +1,122 @@
+/*
+ * Copyright (C) 2023 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License
+ */
+
+package org.conscrypt;
+
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.security.PrivateKey;
+import java.security.PublicKey;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Duck typed implementation of {@link HpkeSpi}.
+ * <p>
+ * Will wrap any Object which implements all of the methods in HpkeSpi and delegate to them
+ * by reflection.
+ */
+@Internal
+public class DuckTypedHpkeSpi implements HpkeSpi {
+  private final Object delegate;
+  private final Map<String, Method> methods = new HashMap<>();
+
+  private DuckTypedHpkeSpi(Object delegate) throws NoSuchMethodException {
+    this.delegate = delegate;
+
+    for (Method targetMethod : HpkeSpi.class.getMethods()) {
+      if (targetMethod.isSynthetic()) {
+        continue;
+      }
+      Class<?> sourceClass = delegate.getClass();
+
+      Method sourceMethod =
+          sourceClass.getMethod(targetMethod.getName(), targetMethod.getParameterTypes());
+      // Check that the return type of obj's method matches the target method.
+      Class<?> sourceReturnType = sourceMethod.getReturnType();
+      Class<?> targetReturnType = targetMethod.getReturnType();
+      if (!targetReturnType.isAssignableFrom(sourceReturnType)) {
+        throw new NoSuchMethodException(sourceMethod + " return value (" + sourceReturnType
+            + ") incompatible with target return value (" + targetReturnType + ")");
+      }
+      methods.put(sourceMethod.getName(), sourceMethod);
```

Comment:
nit: I don't know how likely it is, but if a method from `HpkeSpi` is overloaded, this will be overridden. Should we check that `put` returns null?

Reply:
I think that's covered, as we do
```
 Method sourceMethod =
          sourceClass.getMethod(targetMethod.getName(), targetMethod.getParameterTypes());
```
and (as I found during refactorings :) that won't match unless the parameters match exactly.

I _am_ a bit concerned about differing checked exceptions though....

Also, I'm going to move that `getClass()` outside the loop though seeing as there's another fix needs to go in